pkg/syncer/specsyncer.go

```diff
@@ -143,11 +175,44 @@ func (c *Controller) ensureDownstreamNamespaceExists(ctx context.Context, downst
 	return nil
 }
 
-func (c *Controller) applyToDownstream(ctx context.Context, gvr schema.GroupVersionResource, downstreamNamespace string, upstreamObj *unstructured.Unstructured) error {
+func (c *Controller) notifySyncerOwnership(ctx context.Context, gvr schema.GroupVersionResource, upstreamObj *unstructured.Unstructured) error {
+	name := upstreamObj.GetName()
+	namespace := upstreamObj.GetNamespace()
+
+	// If the advanced scheduling feature is enabled, add the Syncer Finalizer to the upstream object
+	if advancedSchedulingFeatureEnabled {
+		upstreamFinalizers := upstreamObj.GetFinalizers()
+		hasFinalizer := false
+		for _, finalizer := range upstreamFinalizers {
+			if finalizer == SyncerFinalizerName(c.pcluster) {
+				hasFinalizer = true
+			}
+		}
+		if !hasFinalizer {
+			upstreamFinalizers = append(upstreamFinalizers, SyncerFinalizerName(c.pcluster))
+			upstreamObj.SetFinalizers(upstreamFinalizers)
+		}
+	}
+
+	if _, err := c.fromClient.Resource(gvr).Namespace(namespace).Update(ctx, upstreamObj, metav1.UpdateOptions{}); err != nil {
+		klog.Errorf("Failed updating to notify syncer ownership on resource %s|%s/%s: %v", c.upstreamClusterName, namespace, name, err)
+		return err
+	}
+	klog.Infof("Updated resource %s|%s/%s to notify syncer ownership", c.upstreamClusterName, namespace, name)
+	return nil
+}
+
+func (c *Controller) applyToDownstream(ctx context.Context, eventType watch.EventType, gvr schema.GroupVersionResource, downstreamNamespace string, upstreamObj *unstructured.Unstructured) error {
 	if err := c.ensureDownstreamNamespaceExists(ctx, downstreamNamespace, upstreamObj); err != nil {
 		return err
 	}
 
+	if eventType == watch.Added {
```

Comment:
don't act by edges, but by level (= state). Just check whether the finalizer is there and update if it is not.

Reply:
done